### ISSUE TITLE
[Bug] Exporting .anim without previewing result in odd files

### ISSUE THREAD

**Issue by AceyT** [jam1garner/Smash-Forge#56]
### Just a warning for future issues / questions from the community :

~~Exporting .anim files in a Debug build of the software can cause a discrepancy between the actual and the resulted animation~~ -> You have to preview it

For example, the animation of the sword of Link during his first neutral A have only minor changes in the key-frame data, however, only the first two key are properly write onto the file in Debug, whereas everything seems correct once compiled in a Release mode.

I have figured it out by analyzing the code in a line by line execution with the debugger. The variables where set properly, and so the writing was ok as long as I stayed in the _writeKey_ function in ANIM.cs.
As soon as I stepped out in the middle of the loop for frames data, I suppose Windows have "optimized" the execution of the function in some way.

Voilà ! ~~So do not hesitate to tell people to use the tool with the Release version (other tools sometimes precised it, and I didn't know that it would that much of a deal)~~ -> Do not hesitate to tell people to preview all of their animations.

## Edit

Actually, they can still be "optimized" on Release configuration but I don't know why, and it keep getting weirder as some previous export that were fine on Debug aren't anymore on Release configuration.

---

Debug output (if not watched) :

> anim translate.translateX translateX KenJ 0 0 0;
animData {
 input time;
 output linear;
 weighted 1;
 preInfinity constant;
 postInfinity constant;
 keys {
 1 0.658698 fixed fixed 1 1 0 0 1 0 1;
 2 0.658698 fixed fixed 1 1 0 0 1 0 1;
 3 0.000000 fixed fixed 1 1 0 0 1 0 1;
 4 0.000000 fixed fixed 1 1 0 0 1 0 1;
 5 0.000000 fixed fixed 1 1 0 0 1 0 1;
 6 0.000000 fixed fixed 1 1 0 0 1 0 1;
 7 0.000000 fixed fixed 1 1 0 0 1 0 1;
 8 0.000000 fixed fixed 1 1 0 0 1 0 1;
 9 0.000000 fixed fixed 1 1 0 0 1 0 1;
 10 0.000000 fixed fixed 1 1 0 0 1 0 1;
 11 0.000000 fixed fixed 1 1 0 0 1 0 1;
 12 0.000000 fixed fixed 1 1 0 0 1 0 1;
 13 0.000000 fixed fixed 1 1 0 0 1 0 1;
 14 0.000000 fixed fixed 1 1 0 0 1 0 1;
 15 0.000000 fixed fixed 1 1 0 0 1 0 1;
 16 0.000000 fixed fixed 1 1 0 0 1 0 1;
 17 0.000000 fixed fixed 1 1 0 0 1 0 1;
 18 0.000000 fixed fixed 1 1 0 0 1 0 1;
 19 0.000000 fixed fixed 1 1 0 0 1 0 1;
 20 0.000000 fixed fixed 1 1 0 0 1 0 1;
 21 0.000000 fixed fixed 1 1 0 0 1 0 1;
 22 0.000000 fixed fixed 1 1 0 0 1 0 1;
 23 0.000000 fixed fixed 1 1 0 0 1 0 1;
 24 0.000000 fixed fixed 1 1 0 0 1 0 1;
 25 0.000000 fixed fixed 1 1 0 0 1 0 1;
 26 0.000000 fixed fixed 1 1 0 0 1 0 1;
 27 0.000000 fixed fixed 1 1 0 0 1 0 1;
 28 0.000000 fixed fixed 1 1 0 0 1 0 1;
 }
}

Release output :

> anim translate.translateX translateX KenJ 0 0 0;
animData {
 input time;
 output linear;
 weighted 1;
 preInfinity constant;
 postInfinity constant;
 keys {
 1 0.658698 fixed fixed 1 1 0 0 1 0 1;
 2 0.658698 fixed fixed 1 1 0 0 1 0 1;
 3 0.658698 fixed fixed 1 1 0 0 1 0 1;
 4 0.658698 fixed fixed 1 1 0 0 1 0 1;
 5 0.658698 fixed fixed 1 1 0 0 1 0 1;
 6 1.127548 fixed fixed 1 1 0 0 1 0 1;
 7 0.658698 fixed fixed 1 1 0 0 1 0 1;
 8 0.658698 fixed fixed 1 1 0 0 1 0 1;
 9 0.658698 fixed fixed 1 1 0 0 1 0 1;
 10 0.658698 fixed fixed 1 1 0 0 1 0 1;
 11 0.658698 fixed fixed 1 1 0 0 1 0 1;
 12 0.658698 fixed fixed 1 1 0 0 1 0 1;
 13 0.658698 fixed fixed 1 1 0 0 1 0 1;
 14 0.658698 fixed fixed 1 1 0 0 1 0 1;
 15 0.658698 fixed fixed 1 1 0 0 1 0 1;
 16 0.658698 fixed fixed 1 1 0 0 1 0 1;
 17 0.658698 fixed fixed 1 1 0 0 1 0 1;
 18 0.658698 fixed fixed 1 1 0 0 1 0 1;
 19 0.658698 fixed fixed 1 1 0 0 1 0 1;
 20 0.658698 fixed fixed 1 1 0 0 1 0 1;
 21 0.658698 fixed fixed 1 1 0 0 1 0 1;
 22 0.658698 fixed fixed 1 1 0 0 1 0 1;
 23 0.658698 fixed fixed 1 1 0 0 1 0 1;
 24 0.658698 fixed fixed 1 1 0 0 1 0 1;
 25 0.658698 fixed fixed 1 1 0 0 1 0 1;
 26 0.658698 fixed fixed 1 1 0 0 1 0 1;
 27 0.658698 fixed fixed 1 1 0 0 1 0 1;
 28 0.658698 fixed fixed 1 1 0 0 1 0 1;
 }
}

**Comment by AceyT** [comment]
Okay, so at first I thought it was because of overwriting or not a file already existing, but that's not it.

Rather, **if you don't have previewed the animation**, the bug will appear, so it seems to be important to view at least once the entire animation, then export it.

**Comment by AceyT** [comment]
So, I've found a difference in the export function of a new .omo and a of new .anim.
From my test, it's what has been the difference between a correct and a bad .anim export.

In the function ```OMO.createOMO```, it goes on each frame through ```SkelAnimation.nextFrame(VBN)```, and this line seems to do a bunch of update on the VBN linked to the SkelAnimation.

So I've added a bunch of lines in ```ANIM.createANIM```, just before the function get the node indexes from the animation, and here what I'm doing :
~~~ C#
 a.setFrame( a.size() - 1 ); //from last frame
 for (int li = 0 ; li < a.size() ; ++li ) //go through each frame with nextFrame
     a.nextFrame(vbn);
 a.nextFrame(vbn);  //go on first frame

List<int> nodes = a.getNodes (true, vbn); //getting node indexes
//...
~~~

Maybe it isn't optimized, or not in accordance with some coding style that you have, so I'll leave it here so that you'll be able to correct the bug with your way of coding.

Thanks for your work so far, I can now export all the animation to Maya easily without problem.
